lib/kennel/syncer.rb

```diff
@@ -130,45 +130,56 @@ def calculate_changes
         populate_id_map @expected, @actual
         filter_actual! @actual
         resolve_linked_tracking_ids! @expected # resolve dependencies to avoid diff
+        @expected.each(&:add_tracking_id) # avoid diff with actual, which has tracking_id
 
-        @expected.each(&:add_tracking_id) # avoid diff with actual
+        # see which expected match the actual
+        matching, unmatched_expected, unmatched_actual = partition_matched_expected
+        validate_expected_id_not_missing unmatched_expected
+        fill_details! matching # need details to diff later
 
-        lookup_map = matching_expected_lookup_map
-        items = @actual.map do |a|
-          e = matching_expected(a, lookup_map)
-          if e && @expected.delete?(e)
-            [e, a]
-          else
-            [nil, a]
-          end
+        # update matching if needed
+        matching.each do |e, a|
+          id = a.fetch(:id)
+          diff = e.diff(a)
+          @update << [id, e, a, diff] if diff.any?
         end
```

Comment:
```suggestion
        @update = matching.map do |e, a|
          id = a.fetch(:id)
          diff = e.diff(a)
          [id, e, a, diff] if diff.any?
        end.compact
```

Does that work?

Reply:
ah yes ✨ 
`filter_map` ideally once we are past 2.6